process/sync/shardForkDetector_test.go

```diff
@@ -0,0 +1,178 @@
+	assert.Nil(t, sfd)
+	assert.Equal(t, process.ErrNilBlackListHandler, err)
+}
+
+func TestNewShardForkDetector_OkParamsShouldWork(t *testing.T) {
+	t.Parallel()
+
+	sfd, err := sync.NewShardForkDetector(&mock.RounderMock{}, &mock.BlackListHandlerStub{})
+	assert.Nil(t, err)
+	assert.NotNil(t, sfd)
+
+	assert.Equal(t, uint64(0), sfd.LastCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.LastCheckpointRound())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointNonce())
+	assert.Equal(t, uint64(0), sfd.FinalCheckpointRound())
+}
+
+func TestShardForkDetector_AddHeaderNilHeaderShouldErr(t *testing.T) {
+	t.Parallel()
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(nil, make([]byte, 0), process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHeader, err)
+}
+
+func TestShardForkDetector_AddHeaderNilHashShouldErr(t *testing.T) {
+	t.Parallel()
+	rounderMock := &mock.RounderMock{RoundIndex: 100}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{})
+	err := bfd.AddHeader(&block.Header{}, nil, process.BHProcessed, nil, nil, false)
+	assert.Equal(t, sync.ErrNilHash, err)
+}
+
+func TestShardForkDetector_AddHeaderUnsignedBlockShouldErr(t *testing.T) {
+	t.Parallel()
+	rounderMock := &mock.RounderMock{RoundIndex: 1}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{
+		AddCalled: func(key string) error {
+			return nil
+		},
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+	err := bfd.AddHeader(
+		&block.Header{Nonce: 1, Round: 1},
+		make([]byte, 0),
+		process.BHProcessed,
+		nil,
+		nil,
+		false)
+	assert.Equal(t, sync.ErrBlockIsNotSigned, err)
+}
+
+func TestShardForkDetector_AddHeaderNotPresentShouldWork(t *testing.T) {
+	t.Parallel()
+	hdr := &block.Header{Nonce: 1, Round: 1, PubKeysBitmap: []byte("X")}
+	hash := make([]byte, 0)
+	rounderMock := &mock.RounderMock{RoundIndex: 1}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+
+	err := bfd.AddHeader(hdr, hash, process.BHProcessed, nil, nil, false)
+	assert.Nil(t, err)
+
+	hInfos := bfd.GetHeaders(1)
+	assert.Equal(t, 1, len(hInfos))
+	assert.Equal(t, hash, hInfos[0].Hash())
+}
+
+func TestShardForkDetector_AddHeaderPresentShouldAppend(t *testing.T) {
+	t.Parallel()
+	hdr1 := &block.Header{Nonce: 1, Round: 1, PubKeysBitmap: []byte("X")}
+	hash1 := []byte("hash1")
+	hdr2 := &block.Header{Nonce: 1, Round: 1, PubKeysBitmap: []byte("X")}
+	hash2 := []byte("hash2")
+	rounderMock := &mock.RounderMock{RoundIndex: 1}
+	bfd, _ := sync.NewShardForkDetector(rounderMock, &mock.BlackListHandlerStub{
+		HasCalled: func(key string) bool {
+			return false
+		},
+	})
+
+	_ = bfd.AddHeader(hdr1, hash1, process.BHProcessed, nil, nil, false)
+	err := bfd.AddHeader(hdr2, hash2, process.BHProcessed, nil, nil, false)
```

Comment:
also here. why do you check only the second time for error?

Reply:
Same answer as before.